### PR TITLE
Start refactoring the way we generate Option types, add function/callback support to LuaJIT FFI Gen

### PR DESCRIPTION
**Pull request by dgavedissian**
I would recommend reviewing commit by commit. I tried to add comprehensive test coverage to the callback types. I'll try to retrospectively fill in the missing test coverage to all other types over time, although we do still have pretty decent coverage (and there's no change to the generated lua files).

This is what the C types look like:

```
    do -- C Definitions
        ffi.cdef [[
            void   CallbackTest_Free                  (CallbackTest*);
            void   CallbackTest_NthPrimitive          (CallbackTest const*, uint64 index, void (*)(float));
            void   CallbackTest_NthPrimitiveRef       (CallbackTest const*, uint64 index, void (*)(float));
            void   CallbackTest_NthPrimitiveMut       (CallbackTest*, uint64 index, void (*)(float*));
            void   CallbackTest_NthPrimitiveValOpt    (CallbackTest const*, uint64 index, void (*)(float const*));
            void   CallbackTest_NthPrimitiveRefOpt    (CallbackTest const*, uint64 index, void (*)(float const*));
            void   CallbackTest_NthPrimitiveMutOpt    (CallbackTest*, uint64 index, void (*)(float*));
            void   CallbackTest_InsertPrimitive       (CallbackTest*, float (*)());
            void   CallbackTest_NthNoncopyable        (CallbackTest const*, uint64 index, void (*)(Data*));
            void   CallbackTest_NthNoncopyableRef     (CallbackTest const*, uint64 index, void (*)(Data const*));
            void   CallbackTest_NthNoncopyableMut     (CallbackTest*, uint64 index, void (*)(Data*));
            void   CallbackTest_NthNoncopyableValOpt  (CallbackTest const*, uint64 index, void (*)(Data const*));
            void   CallbackTest_NthNoncopyableRefOpt  (CallbackTest const*, uint64 index, void (*)(Data const*));
            void   CallbackTest_NthNoncopyableMutOpt  (CallbackTest*, uint64 index, void (*)(Data*));
            void   CallbackTest_InsertNoncopyable     (CallbackTest*, Data* (*)());
            void   CallbackTest_ReadPrimitiveArray    (CallbackTest const*, void (*)(float const*, uint64));
            void   CallbackTest_LockPrimitiveArray    (CallbackTest*, void (*)(float*, uint64));
            void   CallbackTest_ReadNoncopyableArray  (CallbackTest const*, void (*)(Data const*, uint64));
            void   CallbackTest_LockNoncopyableArray  (CallbackTest*, void (*)(Data*, uint64));
            cstr   CallbackTest_TransformString       (cstr s, cstr (*)(cstr));
            cstr   CallbackTest_TransformStr          (cstr s, cstr (*)(cstr));
            void   CallbackTest_GetMultipleAndReplace (CallbackTest*, uint64 index, float (*)(float, Data const*));
            uint32 CallbackTest_Passthrough           (uint32 input, uint32 (*)(uint32));
        ]]
    end
```

### CODE REVIEW

File: engine/lib/luajit-ffi-gen/src/impl_item/generate_rust.rs [hunk 1]
```diff
@@ -80,7 +80,7 @@ impl ImplInfo {
         quote! {
             #[no_mangle]
             pub unsafe extern "C" fn #func_ident(#self_token #(#param_tokens),*) #ret_token {
-                tracing::trace!("Calling: {}", #func_name);
+                // tracing::trace!("Calling: {}", #func_name);
```
Comment (HaronK) [inline engine/lib/luajit-ffi-gen/src/impl_item/generate_rust.rs:83]: Do you still need it?

File: engine/lib/luajit-ffi-gen/src/impl_item/generate_rust.rs [hunk 2]
```diff
@@ -262,48 +300,116 @@ impl ImplInfo {
...
+                    .collect();
+
+                // Generate the tokens for the argument list.
+                let arg_tokens: Vec<_> = args
+                    .iter()
+                    .map(|(name, _)| format_ident!("{}", name))
+                    .collect();
+
+                // Convert the Rust expressions to FFI.
+                let mut prelude: Vec<TokenStream> = vec![];
```
Comment (HaronK) [inline engine/lib/luajit-ffi-gen/src/impl_item/generate_rust.rs:344]: I think you don't need `Vec<TokenStream>` here, compiler should deduct it by itself.

File: engine/lib/luajit-ffi-gen/src/impl_item/parse.rs [hunk 3]
```diff
@@ -109,12 +117,21 @@ fn parse_method_attrs(attrs: &mut Vec<Attribute>) -> Result<(Vec<String>, BindAr
     }
 }
 
-fn parse_params<'a>(
+fn parse_params<'a, 'b>(
```
Comment (HaronK) [inline engine/lib/luajit-ffi-gen/src/impl_item/parse.rs:120]: Do we really need a second lifetime? Should not params and generics live in the same scope?

File: engine/lib/luajit-ffi-gen/src/impl_item/parse.rs [hunk 4]
```diff
@@ -218,9 +235,37 @@ fn parse_type(ty: &Type) -> Result<TypeInfo> {
...
+                                    });
+                                }
+                            }
+                        }
+                    }
+                }
             }
 
-            let variant = TypeVariant::from_rust_ffi_str(&type_name);
+            let variant: Option<TypeVariant> = TypeVariant::from_rust_ffi_str(&type_name);
```
Comment (HaronK) [inline engine/lib/luajit-ffi-gen/src/impl_item/parse.rs:268]: `Option<TypeVariant>` should not be needed

File: engine/lib/luajit-ffi-gen/tests/test_impl_callbacks.rs [hunk 5]
```diff
@@ -0,0 +1,399 @@
...
+#[derive(Default)]
+pub struct CallbackTest {
+    val_primitives: Vec<f32>,
+    val_noncopyable: Vec<Data>,
+}
+
+// NOTE: remove 'lua_ffi' parameter to see generated Lua file. Do not commit it!!!
+#[luajit_ffi(lua_ffi = false)]
+impl CallbackTest {
+    pub fn nth_primitive<F: FnOnce(f32) -> ()>(&self, index: usize, callback: F) {
```
Comment (HaronK) [inline engine/lib/luajit-ffi-gen/tests/test_impl_callbacks.rs:15]: Is it mandatory to specify `-> ()` or is it just for testing?

**Review by HaronK — COMMENTED** [review]
1. Can/should we support `impl` variant of the callback? I.e. `fn my_funct(f: impl Fn(u32))`.
2. Would be nice to add some documentation about callback support.

File: engine/lib/luajit-ffi-gen/src/impl_item/parse.rs (hunk 3 above)
Comment (dgavedissian) [reply engine/lib/luajit-ffi-gen/src/impl_item/parse.rs:120]: I think VSCode added this by itself, I don't remember adding this...

File: engine/lib/luajit-ffi-gen/tests/test_impl_callbacks.rs (hunk 5 above)
Comment (dgavedissian) [reply engine/lib/luajit-ffi-gen/tests/test_impl_callbacks.rs:15]: Oh no it isnt! I was just being explicit, I'll make sure to mix and match a bit.

File: engine/lib/luajit-ffi-gen/src/impl_item/generate_rust.rs (hunk 1 above)
Comment (dgavedissian) [reply engine/lib/luajit-ffi-gen/src/impl_item/generate_rust.rs:83]: Oh, I didn't mean to comment this out. I think it's still useful to have.

**Comment by dgavedissian** [comment]
> Can/should we support impl variant of the callback? I.e. fn my_funct(f: impl Fn(u32)).

Yes, I meant to add that as well but forgot to. I'll add both that and raw function objects (`fn(T) -> R`).

> Would be nice to add some documentation about callback support.

Good point, I'll add that too.

**Comment by dgavedissian** [comment]
Scratch that, looks like `fn(...) -> ...` wont work.

The reason is because we need to capture the `extern fn(...) -> ...` callback in the closure to adapt it from the C interface to the Rust interface:

```rust
    let __res__ = CallbackTest::passthrough_raw(
        input,
        |arg0| {
            let ret = callback(arg0.as_str()); // we capture 'callback' here unfortunately
            ret
        },
    );
    ```

It's not possible to convert that input param from `extern fn(*const c_uchar) -> u32` to `fn(&str) -> u32` in this case, without also passing the callback function pointer as an argument.

**Review by HaronK — APPROVED** [review]

**Pull request merged**